File: apps/sdkwork-rtc-flutter-mobile/packages/sdkwork_rtc_flutter_mobile_admin_core/lib/src/services/provider_profile_service.dart
import 'package:sdkwork_rtc_backend_sdk/sdkwork_rtc_backend_sdk.dart'
    as generated;

import '../admin_sdk_mapper.dart';
import '../backend_rtc_client.dart';
import '../models/paginated_list_result.dart';
import '../models/provider_profile.dart';

class ProviderProfileService {
  final SdkworkBackendClient _client;

  ProviderProfileService(this._client);

  Future<PaginatedListResult<ProviderProfile>> list({
    String? provider,
    int? page,
    int? pageSize,
    String? cursor,
    String? search,
    String? sort,
  }) async {
    final response = await _client.rtcProviderProfiles.list(
      page,
      pageSize,
      cursor,
      sort,
      search ?? provider,
    );
    return PaginatedListResult(
      items: backendResponseItems(response)
          .map(ProviderProfile.fromJson)
          .toList(),
      nextCursor: backendResponseNextCursor(response),
    );
  }

  Future<ProviderProfile?> get(String id) async {
    final response = await _client.rtcProviderProfiles.retrieve(id);
    final data = backendResponseData(response);
    return data == null ? null : ProviderProfile.fromJson(data);
  }

  Future<ProviderProfile?> create(ProviderProfileCommand command) async {
    final response = await _client.rtcProviderProfiles.create(
      generated.RtcProviderProfileCommand.fromJson(command.toJson()),
    );
    final data = backendResponseData(response);
    return data == null ? null : ProviderProfile.fromJson(data);
  }

  Future<ProviderProfile?> update(
    String id,
    ProviderProfileCommand command,
  ) async {
    final response = await _client.rtcProviderProfiles.update(
      id,
      generated.RtcProviderProfileCommand.fromJson(command.toJson()),
    );
    final data = backendResponseData(response);
    return data == null ? null : ProviderProfile.fromJson(data);
  }

  Future<ProviderProfile?> disable(String id, {String? reason}) async {
    final response = await _client.rtcProviderProfiles.disable(
      id,
      generated.RtcProviderProfileDisableRequest(reason: reason),
    );
    final data = backendResponseData(response);
    return data == null ? null : ProviderProfile.fromJson(data);
  }

  Future<Map<String, dynamic>?> verify(
    String id,
    String queryKind, {
    int? timeoutMs,
  }) async {
    final response = await _client.rtcProviderProfiles.verify(
      id,
      generated.RtcProviderProfileVerifyRequest(
        queryKind: queryKind,
        timeoutMs: timeoutMs,
      ),
    );
    return backendResponseData(response);
  }

  Future<ProviderProfile?> configureCapabilities(
    String id,
    List<String> enabledCapabilities,
    List<String> disabledCapabilities,
  ) async {
    await _client.rtcProviderProfiles.capabilitiesConfigure(
      id,
      {
        'enabledCapabilities': enabledCapabilities,
        'disabledCapabilities': disabledCapabilities,
      },
    );
    return get(id);
  }
}

File: apps/sdkwork-rtc-h5/packages/sdkwork-rtc-h5-admin-core/src/services/providerProfileService.ts
import type { AuthTokenManager } from "@sdkwork/sdk-common";

import type { ProviderProfile, ProviderProfileCommand } from "../types/providerProfile";
import { resolveBackendRtcClient, type RtcBackendClientOptions, type RtcBackendClientSource } from "./backendClient";

interface ListResponse {
  items: ProviderProfile[];
  nextCursor?: string | null;
}

export class ProviderProfileService {
  private readonly client;

  constructor(
    baseUrlOrClient: RtcBackendClientSource,
    tokenManagerOrOptions?: AuthTokenManager | RtcBackendClientOptions,
  ) {
    this.client = resolveBackendRtcClient(baseUrlOrClient, tokenManagerOrOptions);
  }

  async list(params?: {
    provider?: string;
    page?: number;
    limit?: number;
    cursor?: string;
    search?: string;
    sort?: string;
  }): Promise<ListResponse> {
    const response = await this.client.rtcProviderProfiles.rtc.providerProfiles.list({
      page: params?.page,
      pageSize: params?.limit,
      cursor: params?.cursor,
      q: params?.search,
      sort: params?.sort,
    });
    // The generated SDK unwraps the sdkwork-v3 envelope, so the response is
    // already the `data` payload ({ items, pageInfo }).
    const nextCursor = response.pageInfo?.nextCursor;
    return {
      items: response.items,
      nextCursor: nextCursor && nextCursor.length > 0 ? nextCursor : null,
    };
  }

  async get(id: string): Promise<ProviderProfile> {
    const response = await this.client.rtcProviderProfiles.rtc.providerProfiles.retrieve(id);
    if (!response) {
      throw new Error(`RTC provider profile not found: ${id}`);
    }
    return response;
  }

  async create(command: ProviderProfileCommand): Promise<ProviderProfile> {
    const response = await this.client.rtcProviderProfiles.rtc.providerProfiles.create(
      command as Parameters<
        typeof this.client.rtcProviderProfiles.rtc.providerProfiles.create
      >[0],
    );
    if (!response) {
      throw new Error("Invalid response: missing provider profile data");
    }
    return response;
  }

  async update(id: string, command: ProviderProfileCommand): Promise<ProviderProfile> {
    const response = await this.client.rtcProviderProfiles.rtc.providerProfiles.update(
      id,
      command as Parameters<
        typeof this.client.rtcProviderProfiles.rtc.providerProfiles.update
      >[1],
    );
    if (!response) {
      throw new Error("Invalid response: missing provider profile data");
    }
    return response;
  }

  async disable(id: string, reason?: string): Promise<ProviderProfile> {
    const response = await this.client.rtcProviderProfiles.rtc.providerProfiles.disable(id, {
      reason,
    });
    if (!response) {
      throw new Error("Invalid response: missing provider profile data");
    }
    return response;
  }

  async verify(id: string, queryKind: string, timeoutMs?: number): Promise<unknown> {
    const response = await this.client.rtcProviderProfiles.rtc.providerProfiles.verify(id, {
      queryKind: queryKind as Parameters<
        typeof this.client.rtcProviderProfiles.rtc.providerProfiles.verify
      >[1]["queryKind"],
      timeoutMs,
    });
    return response;
  }

  async configureCapabilities(
    id: string,
    enabledCapabilities: string[],
    disabledCapabilities: string[],
  ): Promise<ProviderProfile> {
    const response = await this.client.rtcProviderProfiles.rtc.providerProfiles.capabilities.configure(
      id,
      {
        enabledCapabilities,
        disabledCapabilities,
      },
    );
    return response;
  }
}
